File: CSharp/ILinkai.Weixin.Sdk/Messaging/MessageProcessor.cs
using Microsoft.Extensions.Logging;

namespace ILinkai.Weixin.Sdk.Messaging;

/// <summary>
/// 消息上下文信息
/// 从微信消息转换而来的统一消息格式
/// </summary>
public class MessageContext
{
    /// <summary>
    /// 消息体（文本内容）
    /// </summary>
    public string Body { get; set; } = string.Empty;

    /// <summary>
    /// 发送者ID
    /// </summary>
    public string From { get; set; } = string.Empty;

    /// <summary>
    /// 接收者ID
    /// </summary>
    public string To { get; set; } = string.Empty;

    /// <summary>
    /// 账户ID
    /// </summary>
    public string AccountId { get; set; } = string.Empty;

    /// <summary>
    /// 消息ID
    /// </summary>
    public string MessageId { get; set; } = string.Empty;

    /// <summary>
    /// 时间戳
    /// </summary>
    public long? Timestamp { get; set; }

    /// <summary>
    /// 上下文令牌
    /// </summary>
    public string? ContextToken { get; set; }

    /// <summary>
    /// 媒体文件路径（已下载和解密）
    /// </summary>
    public string? MediaPath { get; set; }

    /// <summary>
    /// 媒体类型
    /// </summary>
    public string? MediaType { get; set; }

    /// <summary>
    /// 原始消息
    /// </summary>
    public Models.WeixinMessage? OriginalMessage { get; set; }
}

/// <summary>
/// 消息处理服务
/// 处理接收到的微信消息
/// </summary>
public class MessageProcessor
{
    private readonly Media.MediaDownloadService _downloadService;
    private readonly ILogger? _logger;

    /// <summary>
    /// 创建MessageProcessor实例
    /// </summary>
    /// <param name="cdnBaseUrl">CDN基础URL</param>
    /// <param name="logger">日志记录器</param>
    public MessageProcessor(string cdnBaseUrl, ILogger? logger = null)
    {
        _downloadService = new Media.MediaDownloadService(cdnBaseUrl, logger);
        _logger = logger;
    }

    /// <summary>
    /// 将微信消息转换为消息上下文
    /// </summary>
    /// <param name="message">微信消息</param>
    /// <param name="accountId">账户ID</param>
    /// <returns>消息上下文</returns>
    public MessageContext ConvertToContext(Models.WeixinMessage message, string accountId)
    {
        var fromUserId = message.FromUserId ?? "";
        var body = ExtractBodyFromItemList(message.ItemList);

        return new MessageContext
        {
            Body = body,
            From = fromUserId,
            To = fromUserId,
            AccountId = accountId,
            MessageId = $"ilinkai-{Guid.NewGuid():N}",
            Timestamp = message.CreateTimeMs,
            ContextToken = message.ContextToken,
            OriginalMessage = message
        };
    }

    /// <summary>
    /// 处理消息中的媒体内容
    /// </summary>
    /// <param name="context">消息上下文</param>
    /// <param name="mediaDir">媒体存储目录</param>
    /// <param name="cancellationToken">取消令牌</param>
    public async Task ProcessMediaAsync(MessageContext context, string mediaDir, CancellationToken cancellationToken = default)
    {
        var message = context.OriginalMessage;
        if (message?.ItemList == null || message.ItemList.Count == 0)
        {
            return;
        }

        foreach (var item in message.ItemList)
        {
            if (item.Type == null) continue;

            try
            {
                switch ((Models.MessageItemType)item.Type)
                {
                    case Models.MessageItemType.Image:
                        await ProcessImageItemAsync(context, item, mediaDir, cancellationToken);
                        break;

                    case Models.MessageItemType.Voice:
                        await ProcessVoiceItemAsync(context, item, mediaDir, cancellationToken);
                        break;

                    case Models.MessageItemType.File:
                        await ProcessFileItemAsync(context, item, mediaDir, cancellationToken);
                        break;

                    case Models.MessageItemType.Video:
                        await ProcessVideoItemAsync(context, item, mediaDir, cancellationToken);
                        break;
                }

                if (!string.IsNullOrEmpty(context.MediaPath))
                {
                    break;
                }
            }
            catch (Exception err)
            {
                _logger?.LogError(err, "ProcessMedia: failed to process item type={Type}", item.Type);
            }
        }
    }

    /// <summary>
    /// 处理图片项
    /// </summary>
    private async Task ProcessImageItemAsync(MessageContext context, Models.MessageItem item, string mediaDir, CancellationToken cancellationToken)
    {
        var img = item.ImageItem;
        if (img?.Media?.EncryptQueryParam == null) return;

        var aesKeyBase64 = !string.IsNullOrEmpty(img.AesKey)
            ? Convert.ToBase64String(Convert.FromHexString(img.AesKey))
            : img.Media.AesKey;

        _logger?.LogDebug("ProcessImage: hasAesKey={HasKey}", !string.IsNullOrEmpty(aesKeyBase64));

        try
        {
            var data = await _downloadService.DownloadImageAsync(img.Media.EncryptQueryParam, aesKeyBase64, cancellationToken);
            context.MediaPath = await _downloadService.SaveMediaToFileAsync(data, mediaDir, $"image_{DateTime.UtcNow:yyyyMMdd_HHmmss}.jpg", cancellationToken);
            context.MediaType = "image/*";
            _logger?.LogDebug("ProcessImage: saved to {Path}", context.MediaPath);
        }
        catch (Exception err)
        {
            _logger?.LogError(err, "ProcessImage: download/decrypt failed");
        }
    }

    /// <summary>
    /// 处理语音项
    /// </summary>
    private async Task ProcessVoiceItemAsync(MessageContext context, Models.MessageItem item, string mediaDir, CancellationToken cancellationToken)
    {
        var voice = item.VoiceItem;
        if (voice?.Media?.EncryptQueryParam == null || string.IsNullOrEmpty(voice.Media.AesKey)) return;

        try
        {
            var data = await _downloadService.DownloadVoiceAsync(voice.Media.EncryptQueryParam, voice.Media.AesKey, cancellationToken);
            context.MediaPath = await _downloadService.SaveMediaToFileAsync(data, mediaDir, $"voice_{DateTime.UtcNow:yyyyMMdd_HHmmss}.silk", cancellationToken);
            context.MediaType = "audio/silk";
            _logger?.LogDebug("ProcessVoice: saved to {Path}", context.MediaPath);
        }
        catch (Exception err)
        {
            _logger?.LogError(err, "ProcessVoice: download failed");
        }
    }

    /// <summary>
    /// 处理文件项
    /// </summary>
    private async Task ProcessFileItemAsync(MessageContext context, Models.MessageItem item, string mediaDir, CancellationToken cancellationToken)
    {
        var fileItem = item.FileItem;
        if (fileItem?.Media?.EncryptQueryParam == null || string.IsNullOrEmpty(fileItem.Media.AesKey)) return;

        try
        {
            var data = await _downloadService.DownloadFileAsync(fileItem.Media.EncryptQueryParam, fileItem.Media.AesKey, cancellationToken);
            var fileName = !string.IsNullOrEmpty(fileItem.FileName) ? fileItem.FileName : $"file_{DateTime.UtcNow:yyyyMMdd_HHmmss}.bin";
            context.MediaPath = await _downloadService.SaveMediaToFileAsync(data, mediaDir, fileName, cancellationToken);
            context.MediaType = GetMimeTypeFromFileName(fileName);
            _logger?.LogDebug("ProcessFile: saved to {Path}", context.MediaPath);
        }
        catch (Exception err)
        {
            _logger?.LogError(err, "ProcessFile: download failed");
        }
    }

    /// <summary>
    /// 处理视频项
    /// </summary>
    private async Task ProcessVideoItemAsync(MessageContext context, Models.MessageItem item, string mediaDir, CancellationToken cancellationToken)
    {
        var videoItem = item.VideoItem;
        if (videoItem?.Media?.EncryptQueryParam == null || string.IsNullOrEmpty(videoItem.Media.AesKey)) return;

        try
        {
            var data = await _downloadService.DownloadVideoAsync(videoItem.Media.EncryptQueryParam, videoItem.Media.AesKey, cancellationToken);
            context.MediaPath = await _downloadService.SaveMediaToFileAsync(data, mediaDir, $"video_{DateTime.UtcNow:yyyyMMdd_HHmmss}.mp4", cancellationToken);
            context.MediaType = "video/mp4";
            _logger?.LogDebug("ProcessVideo: saved to {Path}", context.MediaPath);
        }
        catch (Exception err)
        {
            _logger?.LogError(err, "ProcessVideo: download failed");
        }
    }

    /// <summary>
    /// 从消息项列表提取文本内容
    /// </summary>
    public static string ExtractBodyFromItemList(List<Models.MessageItem>? itemList)
    {
        if (itemList == null || itemList.Count == 0) return string.Empty;

        foreach (var item in itemList)
        {
            if (item.Type == (int)Models.MessageItemType.Text && item.TextItem?.Text != null)
            {
                var text = item.TextItem.Text;
                var refMsg = item.RefMessage;

                if (refMsg == null) return text;

                if (refMsg.MessageItem != null && IsMediaItem(refMsg.MessageItem))
                {
                    return text;
                }

                var parts = new List<string>();
                if (!string.IsNullOrEmpty(refMsg.Title))
                {
                    parts.Add(refMsg.Title);
                }
                if (refMsg.MessageItem != null)
                {
                    var refBody = ExtractBodyFromItemList(new List<Models.MessageItem> { refMsg.MessageItem });
                    if (!string.IsNullOrEmpty(refBody))
                    {
                        parts.Add(refBody);
                    }
                }

                if (parts.Count == 0) return text;
                return $"[引用: {string.Join(" | ", parts)}]\n{text}";
            }

            if (item.Type == (int)Models.MessageItemType.Voice && item.VoiceItem?.Text != null)
            {
                return item.VoiceItem.Text;
            }
        }

        return string.Empty;
    }

    /// <summary>
    /// 检查是否为媒体项
    /// </summary>
    public static bool IsMediaItem(Models.MessageItem item)
    {
        return item.Type == (int)Models.MessageItemType.Image ||
               item.Type == (int)Models.MessageItemType.Video ||
               item.Type == (int)Models.MessageItemType.File ||
               item.Type == (int)Models.MessageItemType.Voice;
    }

    /// <summary>
    /// 根据文件名获取MIME类型
    /// </summary>
    private static string GetMimeTypeFromFileName(string fileName)
    {
        var ext = Path.GetExtension(fileName).ToLowerInvariant();
        return ext switch
        {
            ".jpg" or ".jpeg" => "image/jpeg",
            ".png" => "image/png",
            ".gif" => "image/gif",
            ".pdf" => "application/pdf",
            ".doc" => "application/msword",
            ".docx" => "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
            ".xls" => "application/vnd.ms-excel",
            ".xlsx" => "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            ".zip" => "application/zip",
            ".mp3" => "audio/mpeg",
            ".mp4" => "video/mp4",
            _ => "application/octet-stream"
        };
    }
}

File: Python/ilinkai_weixin/messaging/message_processor.py
"""消息处理服务"""
import base64
import logging
import os
import uuid
from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional, List

from ..media.media_download_service import MediaDownloadService
from ..models import WeixinMessage, MessageItem, MessageItemType

logger = logging.getLogger(__name__)


@dataclass
class MessageContext:
    body: str = ""
    from_: str = ""
    to: str = ""
    account_id: str = ""
    message_id: str = ""
    timestamp: Optional[int] = None
    context_token: Optional[str] = None
    media_path: Optional[str] = None
    media_type: Optional[str] = None
    original_message: Optional[WeixinMessage] = None


class MessageProcessor:
    def __init__(self, cdn_base_url: str):
        self._download_service = MediaDownloadService(cdn_base_url)

    def convert_to_context(self, message: WeixinMessage, account_id: str) -> MessageContext:
        from_user_id = message.from_user_id or ""
        body = extract_body_from_item_list(message.item_list)
        return MessageContext(
            body=body, from_=from_user_id, to=from_user_id,
            account_id=account_id,
            message_id=f"ilinkai-{uuid.uuid4().hex}",
            timestamp=message.create_time_ms,
            context_token=message.context_token,
            original_message=message,
        )

    def process_media(self, context: MessageContext, media_dir: str):
        message = context.original_message
        if not message or not message.item_list:
            return
        for item in message.item_list:
            if item.type is None:
                continue
            try:
                if item.type == MessageItemType.IMAGE:
                    self._process_image_item(context, item, media_dir)
                elif item.type == MessageItemType.VOICE:
                    self._process_voice_item(context, item, media_dir)
                elif item.type == MessageItemType.FILE:
                    self._process_file_item(context, item, media_dir)
                elif item.type == MessageItemType.VIDEO:
                    self._process_video_item(context, item, media_dir)
                if context.media_path:
                    break
            except Exception as err:
                logger.error("ProcessMedia: failed to process item type=%s: %s", item.type, err)

    def _process_image_item(self, ctx: MessageContext, item: MessageItem, media_dir: str):
        img = item.image_item
        if not img or not img.media or not img.media.encrypt_query_param:
            return
        aes_key_base64 = None
        if img.aes_key:
            aes_key_base64 = base64.b64encode(bytes.fromhex(img.aes_key)).decode()
        elif img.media.aes_key:
            aes_key_base64 = img.media.aes_key
        data = self._download_service.download_image(img.media.encrypt_query_param, aes_key_base64)
        ts = datetime.utcnow().strftime('%Y%m%d_%H%M%S')
        ctx.media_path = self._download_service.save_media_to_file(data, media_dir, f"image_{ts}.jpg")
        ctx.media_type = "image/*"

    def _process_voice_item(self, ctx: MessageContext, item: MessageItem, media_dir: str):
        voice = item.voice_item
        if not voice or not voice.media or not voice.media.encrypt_query_param or not voice.media.aes_key:
            return
        data = self._download_service.download_voice(voice.media.encrypt_query_param, voice.media.aes_key)
        ts = datetime.utcnow().strftime('%Y%m%d_%H%M%S')
        ctx.media_path = self._download_service.save_media_to_file(data, media_dir, f"voice_{ts}.silk")
        ctx.media_type = "audio/silk"

    def _process_file_item(self, ctx: MessageContext, item: MessageItem, media_dir: str):
        fi = item.file_item
        if not fi or not fi.media or not fi.media.encrypt_query_param or not fi.media.aes_key:
            return
        data = self._download_service.download_file(fi.media.encrypt_query_param, fi.media.aes_key)
        file_name = fi.file_name or f"file_{datetime.utcnow().strftime('%Y%m%d_%H%M%S')}.bin"
        ctx.media_path = self._download_service.save_media_to_file(data, media_dir, file_name)
        ctx.media_type = _get_mime_type(file_name)

    def _process_video_item(self, ctx: MessageContext, item: MessageItem, media_dir: str):
        vi = item.video_item
        if not vi or not vi.media or not vi.media.encrypt_query_param or not vi.media.aes_key:
            return
        data = self._download_service.download_video(vi.media.encrypt_query_param, vi.media.aes_key)
        ts = datetime.utcnow().strftime('%Y%m%d_%H%M%S')
        ctx.media_path = self._download_service.save_media_to_file(data, media_dir, f"video_{ts}.mp4")
        ctx.media_type = "video/mp4"


def extract_body_from_item_list(item_list: Optional[List[MessageItem]]) -> str:
    if not item_list:
        return ""
    for item in item_list:
        if item.type == MessageItemType.TEXT and item.text_item and item.text_item.text is not None:
            text = item.text_item.text
            ref_msg = item.ref_message
            if not ref_msg:
                return text
            if ref_msg.message_item and is_media_item(ref_msg.message_item):
                return text
            parts = []
            if ref_msg.title:
                parts.append(ref_msg.title)
            if ref_msg.message_item:
                ref_body = extract_body_from_item_list([ref_msg.message_item])
                if ref_body:
                    parts.append(ref_body)
            if not parts:
                return text
            return f"[引用: {' | '.join(parts)}]\n{text}"
        if item.type == MessageItemType.VOICE and item.voice_item and item.voice_item.text is not None:
            return item.voice_item.text
    return ""


def is_media_item(item: MessageItem) -> bool:
    return item.type in (MessageItemType.IMAGE, MessageItemType.VIDEO,
                         MessageItemType.FILE, MessageItemType.VOICE)


_MIME_MAP = {
    '.jpg': 'image/jpeg', '.jpeg': 'image/jpeg', '.png': 'image/png', '.gif': 'image/gif',
    '.pdf': 'application/pdf', '.doc': 'application/msword',
    '.docx': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    '.xls': 'application/vnd.ms-excel',
    '.xlsx': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    '.zip': 'application/zip', '.mp3': 'audio/mpeg', '.mp4': 'video/mp4',
}

def _get_mime_type(file_name: str) -> str:
    _, ext = os.path.splitext(file_name.lower())
    return _MIME_MAP.get(ext, 'application/octet-stream')
